src/normed.jl

```diff
@@ -41,27 +41,61 @@ rawone(v) = reinterpret(one(v))
 function Normed{T,f}(x::Normed{T2}) where {T <: Unsigned,T2 <: Unsigned,f}
     U = Normed{T,f}
     y = round((rawone(U)/rawone(x))*reinterpret(x))
-    (0 <= y) & (y <= typemax(T)) || throw_converterror(U, x)
+    0 <= y <= typemax(T) || throw_converterror(U, x)
```

Comment:
These are actually subtly different, and at least at one point there was a performance difference (doesn't seem to be one now, at least not on my current machine):
```julia
julia> function f1(x)
           0.25 <= x <= 0.75 || throw(DomainError(x))
           return x
       end
f1 (generic function with 1 method)

julia> function f2(x)
           ((0.25 <= x) & (x <= 0.75)) || throw(DomainError(x))
           return x
       end
f2 (generic function with 1 method)

julia> @code_lowered f1(0.5)
CodeInfo(
1 ─ %1 = 0.25 <= x
└──      goto #3 if not %1
2 ─      @_3 = x <= 0.75
└──      goto #4
3 ─      @_3 = false
4 ┄ %6 = @_3
└──      goto #6 if not %6
5 ─      goto #7
6 ─ %9 = Main.DomainError(x)
└──      Main.throw(%9)
7 ┄      return x
)

julia> @code_lowered f2(0.5)
CodeInfo(
1 ─ %1 = 0.25 <= x
│   %2 = x <= 0.75
│   %3 = %1 & %2
└──      goto #3 if not %3
2 ─      goto #4
3 ─ %6 = Main.DomainError(x)
└──      Main.throw(%6)
4 ┄      return x
)

julia> using BenchmarkTools

julia> @btime f1(x) setup=(x=rand()/10 + 0.3)
  1.980 ns (0 allocations: 0 bytes)
0.34079130696459065

julia> @btime f2(x) setup=(x=rand()/10 + 0.3)
  1.977 ns (0 allocations: 0 bytes)
0.30934306198706746
```

Reply:
Yes. I guess, using `&` instead of `&&` may reduce condition branches. However, `throw_converterror(U, x)` is annotated with `@noinline` and throws the exception. This means that we cannot avoid condition branches, and this is not SIMD-suitable. Depending on the type (and CPU), the comparison with zero can be specialized (this is the reason why I did not use `zero()`).

Of course, I follow the law of *"if it works, don't touch it"*.:smile:

Reply:
Yeah, you definitely need one branch, but one is usually better than two. For example when Matt & I were developing the array infrastructure we noticed it was better to check bounds on all indexes and combine with `&`, and then throw a `BoundsError` if necessary, rather than having one branch per dimension.

But this is definitely something that depends on the CPU and its ability to predict branches, so it can even be hard to benchmark. I would generally suspect that fewer branches is a good thing.

Reply:
That's right. Nevertheless, I think it is another good practice to avoid something tricky and leave it to the compiler (LLVM) *now*, because this is *usually* not a pair of integer comparisons.

However, I do never intend to dispute about the comparison manner, which has no definitive answer. Instead I would care about the `/`, but it is another matter.